### DISCUSSION TITLE
:h fold-expr and Vim9.

### DISCUSSION THREAD

**Discussion by ubaldot** [vim/vim#15777] [Q&A] [answered]
By reading `:h fold-expr` I noticed that the fold expression can return the following:

```
- The result is used for the fold level in this way:
  value			meaning 
  0			the line is not in a fold
  1, 2, ..		the line is in a fold with this level
  -1			the fold level is undefined, use the fold level of a
			line before or after this line, whichever is the
			lowest.
  "="			use fold level from the previous line
  "a1", "a2", ..	add one, two, .. to the fold level of the previous
			line, use the result for the current line
  "s1", "s2", ..	subtract one, two, .. from the fold level of the
			previous line, use the result for the next line
  "<1", "<2", ..	a fold with this level ends at this line
  ">1", ">2", ..	a fold with this level starts at this line
```

However, the fold expression is typically a function call (e.g. `&foldexpr=MyFoldExpr()`) and, to the best of my knowledge, in Vim9 a function shall return a well specified type, whereas here it seems that the expression can return both `number` or `string` types.

How to handle that?

**Comment by zeertzjq** [comment, marked as answer]
Returning a string will work for the number cases.

**Comment by dkearns** [comment]
You can also use `any` for the return type.

**Reply by Konfekt** [reply]
Is this a well specified type? Does in Vim9 general the type `string` comprise `integer` and `any` the type `string`?
